File: problem/0070.java
class Solution {
    public int climbStairs(int n) {
        // // mehtod 1, DP, time = O(N), space = O(N)
        // if (n < 4) return n;
        // int[] dp = new int[n];
        // for (int i = 0; i < n; i++) {
        //     if (i < 3) dp[i] = i + 1;
        //     else dp[i] = dp[i - 1] + dp[i - 2];
        // }
        // return dp[n - 1];
        
        // mehtod 2, DP, time = O(N), space = O(1)
        if (n < 4) return n;
        int prevOfPrev = 1, prev = 2, result = 0;
        while (n-- > 2) {
            result = prev + prevOfPrev;
            prevOfPrev = prev;
            prev = result;
        }
        return result;
    }
}

File: problem/0070.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n < 4: return n
        prevOfPrev, prev, current = 1, 2, 3
        while n > 2:
            current = prevOfPrev + prev
            prevOfPrev = prev
            prev = current
            n -= 1
        return current